src/ray/raylet/worker_killing_policy_group_by_owner.cc

```diff
@@ -0,0 +1,183 @@
+// Copyright 2022 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/raylet/worker_killing_policy_group_by_owner.h"
+
+#include <gtest/gtest_prod.h>
+
+#include <boost/container_hash/hash.hpp>
+#include <unordered_map>
+
+#include "absl/container/flat_hash_map.h"
+#include "ray/common/asio/instrumented_io_context.h"
+#include "ray/common/asio/periodical_runner.h"
+#include "ray/raylet/worker.h"
+#include "ray/raylet/worker_killing_policy.h"
+#include "ray/raylet/worker_pool.h"
+
+namespace ray {
+
+namespace raylet {
+
+GroupByOwnerIdWorkerKillingPolicy::GroupByOwnerIdWorkerKillingPolicy() {}
+
+const std::pair<std::shared_ptr<WorkerInterface>, bool>
+GroupByOwnerIdWorkerKillingPolicy::SelectWorkerToKill(
```

Comment:
how often does this run, would rebuilding the group structure every be a concern?

Reply:
1-2 seconds max? I think the # workers is O(num_cpu), so it should be in the ms range worse case for every 1-2 sec?